Higher order

ShrinivasJC opened the pull request:
There have been many changes as listed in the changelog. Please try varying the input options in the examples to ensure that we can catch as many bugs as possible. Please look-out for unexpected behaviors such as the code taking longer than what you would expect.  Please go through  the documentation and mark any typos or incorrect/missing text.

I have already pulled the development branch into the branch to ensure smooth merging. None of the codes should change once the pull request is accepted.

wahls commented:
Nice work! It compiles on Linux (with gcc 10.1) and all tests pass.

I am not yet done checking everything out, but here are already a few comments:

- It would be nice if the new static auxiliary functions get a short comment on top explaining their purpose (like the existing ones).

- If the first sample of a signal is zero, a division by zero might occur [here](https://github.com/FastNFT/FNFT/blob/d8e146a0bc57abe77dabfab5049dbc9ac5535b60/src/fnft_nsep.c#L124).

- There seems to be some code doubling, the static function `signal_effective_from_signal` for example occurs in three different files with nearly identical content. Is there a reason not to define it only once in a separate file?

- I think the code might be easier to understand if some names are changed:
  + `q_effective` -> `q_upsampled`, `q_resampled` or maybe `q_preprocessed`
  + `signal_effective_from_signal` -> `upsample_signal`, `resample_signal` or `preprocess_signal`
  + `D_scale` -> `upsampling_factor` or `resampling_factor`
  + `nse_discretization_D_scale`, etc. -> as above

- These compiler warnings should be fixed:

```
In file included from /home/sander/git/FNFT-higher_order/include/fnft_errwarn.h:32,
                 from /home/sander/git/FNFT-higher_order/include/private/fnft__errwarn.h:30,
                 from /home/sander/git/FNFT-higher_order/include/fnft_nsep.h:31,
                 from /home/sander/git/FNFT-higher_order/src/fnft_nsep.c:25:
/home/sander/git/FNFT-higher_order/src/fnft_nsep.c: In function ‘fnft_nsep’:
/home/sander/git/FNFT-higher_order/include/fnft_numtypes.h:198:22: warning: using complex absolute value function ‘cabs’ when argument is of floating-point type ‘double’ [-Wabsolute-value]
  198 | #define FNFT_CABS(X) cabs(X)
      |                      ^~~~
/home/sander/git/FNFT-higher_order/include/fnft_numtypes.h:268:25: note: in expansion of macro ‘FNFT_CABS’
  268 | #define CABS(X)         FNFT_CABS(X)
      |                         ^~~~~~~~~
/home/sander/git/FNFT-higher_order/src/fnft_nsep.c:126:13: note: in expansion of macro ‘CABS’
  126 |         if (CABS(1-CABS(m_q)) > 1e-2)
      |             ^~~~
```

```
In file included from /home/sander/git/FNFT-higher_order/include/fnft_errwarn.h:32,
                 from /home/sander/git/FNFT-higher_order/include/fnft.h:24,
                 from /home/sander/git/FNFT-higher_order/include/fnft_nse_discretization_t.h:29,
                 from /home/sander/git/FNFT-higher_order/include/private/fnft__nse_discretization.h:32,
                 from /home/sander/git/FNFT-higher_order/include/private/fnft__nse_scatter.h:30,
                 from /home/sander/git/FNFT-higher_order/src/private/fnft__nse_scatter_bound_states.c:23:
/home/sander/git/FNFT-higher_order/src/private/fnft__nse_scatter_bound_states.c: In function ‘fnft__nse_scatter_bound_states’:
/home/sander/git/FNFT-higher_order/include/fnft_numtypes.h:198:22: warning: using complex absolute value function ‘cabs’ when argument is of floating-point type ‘double’ [-Wabsolute-value]
  198 | #define FNFT_CABS(X) cabs(X)
      |                      ^~~~
/home/sander/git/FNFT-higher_order/include/fnft_numtypes.h:268:25: note: in expansion of macro ‘FNFT_CABS’
  268 | #define CABS(X)         FNFT_CABS(X)
      |                         ^~~~~~~~~
/home/sander/git/FNFT-higher_order/src/private/fnft__nse_scatter_bound_states.c:634:23: note: in expansion of macro ‘CABS’
  634 |                 tmp = CABS((0.5*LOG(CABS((PHI2[n]/PSI2[n])/(PHI1[n]/PSI1[n])))));
      |                       ^~~~
```

ShrinivasJC commented:
Thank you for your keen observations and instructive comments. I have tried and addressed all of them. Please let me know if you want me to add comments about some specific functions.

wahls commented:
Great, thanks! A few more comments:

- The function `refine_roots_newton` is also defined twice, once in `fnft_nsev.c` and once in `fnft_nsev_slow.c`. I would propose to keep the one in the "slow" file and make it available via a new private header file `fnft__nsev_slow.h`.

- Could the numerical constants in `nse_preprocess_signal` be defined in a way such that accuracy is kept if `FNFT_REAL` is changed e.g. to quad precision?

- Still regarding the check for the quasi-periodicity in `fnft_nsep`: I know we agreed on this approach in the past, but I now see some new problems.

  1. When the warning is triggered, the user has no obvious way to fix it.
  2. If the first and last sample are zero plus some small noises, then quasi-periodicity with a random phase is detected even when the signal is periodic. No warning would be triggered in that case even though there are large errors. 

  I am beginning to think that it might be better to explicitly specify the phase-shift over one quasi-period instead. (For example, by providing a new parameter phase_shift after T to fnft_nsep.) Any thoughts?

- Should `square_matrix_mult` be moved into `misc`?

wahls commented:
- `square_matrix_mult` is also defined at least twice -> move to misc?

- Follow-up on the previous constants question: Such constants also appear in `akns_scatter_matrix`, `kdv_scatter_matrix`, and `nse_scatter_bound_states`

ShrinivasJC commented:
> Great, thanks! A few more comments:
> 
> * The function `refine_roots_newton` is also defined twice, once in `fnft_nsev.c` and once in `fnft_nsev_slow.c`. I would propose to keep the one in the "slow" file and make it available via a new private header file `fnft__nsev_slow.h`.

Agree with this.

> * Could the numerical constants in `nse_preprocess_signal` be defined in a way such that accuracy is kept if `FNFT_REAL` is changed e.g. to quad precision?

Yes I will change their definitions here and other places where they show-up.

> * Still regarding the check for the quasi-periodicity in `fnft_nsep`: I know we agreed on this approach in the past, but I now see some new problems.
>   
>   1. When the warning is triggered, the user has no obvious way to fix it.
>   2. If the first and last sample are zero plus some small noises, then quasi-periodicity with a random phase is detected even when the signal is periodic. No warning would be triggered in that case even though there are large errors.
>   
>   I am beginning to think that it might be better to explicitly specify the phase-shift over one quasi-period instead. (For example, by providing a new parameter phase_shift after T to fnft_nsep.) Any thoughts?

I think you are right. Making it an explicit input would be better. Should I proceed with making this change?

> * Should `square_matrix_mult` be moved into `misc`?
Can the function be inlined even then? This function has the highest cost in those routines, so not being able to inline them may affect execution times negatively.

wahls commented:
> > Great, thanks! A few more comments:
> > 
> > * The function `refine_roots_newton` is also defined twice, once in `fnft_nsev.c` and once in `fnft_nsev_slow.c`. I would propose to keep the one in the "slow" file and make it available via a new private header file `fnft__nsev_slow.h`.
> 
> Agree with this.
> 
> > * Could the numerical constants in `nse_preprocess_signal` be defined in a way such that accuracy is kept if `FNFT_REAL` is changed e.g. to quad precision?
> 
> Yes I will change their definitions here and other places where they show-up.
> 
> > * Still regarding the check for the quasi-periodicity in `fnft_nsep`: I know we agreed on this approach in the past, but I now see some new problems.
> >   
> >   1. When the warning is triggered, the user has no obvious way to fix it.
> >   2. If the first and last sample are zero plus some small noises, then quasi-periodicity with a random phase is detected even when the signal is periodic. No warning would be triggered in that case even though there are large errors.
> >   
> >   I am beginning to think that it might be better to explicitly specify the phase-shift over one quasi-period instead. (For example, by providing a new parameter phase_shift after T to fnft_nsep.) Any thoughts?
> 
> I think you are right. Making it an explicit input would be better. Should I proceed with making this change?

Yes, please, this one and the others.

> > * Should `square_matrix_mult` be moved into `misc`?
> >   Can the function be inlined even then? This function has the highest cost in those routines, so not being able to inline them may affect execution times negatively.

Normally not, but if inlineing is important there is also the option to put also the function body in the header file (instead of a separate c file) and keep it `inline static`. See `fnft__fft_wrapper.h` for examples.

ShrinivasJC commented:
Noted. I will add it to misc.h then.

ShrinivasJC commented:
> * Could the numerical constants in `nse_preprocess_signal` be defined in a way such that accuracy is kept if `FNFT_REAL` is changed e.g. to quad precision?

I was able to do it for all discretizations except CF6_4. A set of nonlinear equations have to be solved for CF6_4 which I felt is unnecessary at this stage. 

Please check if other changes are as expected.

wahls commented on examples/fnft_nsep_example.c:
```diff
@@ -80,7 +83,7 @@ int main()
     /** Step 3: Call fnft_nsev and check for errors **/
 
-    int ret_code = fnft_nsep(D, q, T, &K, main_spec, &M, aux_spec, NULL, kappa,
+    int ret_code = fnft_nsep(D, q, T, phase_shift, &K, main_spec, &M, aux_spec, NULL, kappa,
```
Here is another left-over `int` that should actually be an `FNFT_INT`.

ShrinivasJC replied:
Fixed.

wahls commented on include/fnft_nsep.h:
```diff
@@ -189,13 +189,16 @@ fnft_nsep_opts_t fnft_nsep_default_opts();
  *  (i.e., \f$ q(t_0), q(t_1), \dots, q(t_{D-1}) \f$)
  * @param[in] T Array of length 2. T[0] is the position in time of the first
- *  sample. T[1] is the beginning of the next period. It should be T[0]<T[1].
+ *  sample. T[2] is the beginning of the next period. (The location of the last
```
`T[2]` should be `T[1]`, also in the next line.

wahls commented on include/private/fnft__misc.h:
```diff
@@ -235,6 +236,66 @@ FNFT_UINT fnft__misc_nextpowerof2(const FNFT_UINT number);
+static inline void fnft__misc_square_matrix_mult(const FNFT_UINT N, FNFT_COMPLEX * const U,
+        FNFT_COMPLEX *const T){
+    FNFT_UINT  c1, c2, c3;
+    FNFT_COMPLEX TM[32] = { 0 }, sum = 0;
```
The size of `TM` is hard-coded, which I guess is for efficiency? For too large `N`, this might lead to undetected memory violations.

I would suggest to either add a check that `N` is not too large and mention the limit in the docs, or, if that really has an impact performance and it fits the further use in FNFT, define one prototype function

 `misc_square_matrix_mult_proto(N, U, T, TM)` 

and then introduced functions like 

`misc_matrix_mult_2x2(N, U, T)`, 
`misc_matrix_mult_3x3(N, U, T)` 

that locally define `TM` of the right size and pass it to the prototype.

ShrinivasJC replied:
Changed to this suggested structure.

wahls commented on include/private/fnft__nsev_slow.h:
```diff
@@ -0,0 +1,73 @@
+ * @param[in] D Number of samples
+ * @param[in] q Array of length D, contains samples \f$ q_n\f$ for \f$n=0,1,\dots,D-1\f$
+ * in ascending order (i.e., \f$ q_0, q_1, \dots, q_{D-1} \f$). The values
+ * should be specifically precalculated based on the chosen discretization.
```
If you are not already familiar with how the preprocessing works, it might be difficult to figure out from this what to do (and when). Maybe replace the sentence "The values ..." with something like "For certain discretizations, the samples have to be preprocessed. See LINK TO SOMEWHERE WHERE THIS IS EXPLAINED for details."?

ShrinivasJC replied:
Link added.

wahls commented on include/private/fnft__nsev_slow.h:
```diff
@@ -0,0 +1,73 @@
+ * in ascending order (i.e., \f$ q_0, q_1, \dots, q_{D-1} \f$). The values
+ * should be specifically precalculated based on the chosen discretization.
+ * @param[in,out] r Array of length D, contains samples \f$ r_n\f$ for \f$n=0,1,\dots,D-1\f$
+ * in ascending order (i.e., \f$ r_0, r_1, \dots, r_{D-1} \f$). The values
```
Same here.

wahls commented on include/private/fnft__nsev_slow.h:
```diff
@@ -0,0 +1,73 @@
+ * the initial guesses for the bound-states \f$\lambda\f$. On return it 
+ * contains the refined bound states.
+ * @param[in] discretization The type of discretization to be used. Should be of type 
+ * \link fnft_nse_discretization_t \endlink. Not all nse_discretization_t discretizations are supported.
```
Mention the supported ones?

ShrinivasJC replied:
Mentioned supported discretizations.

wahls commented on include/private/fnft__misc.h:
```diff
@@ -235,6 +236,66 @@ FNFT_UINT fnft__misc_nextpowerof2(const FNFT_UINT number);
+ * @param[in,out] T Pointer to the first element of complex values NxN matrix T.
+ * Contains the result U*T on return.
+ */
+static inline void fnft__misc_square_matrix_mult(const FNFT_UINT N, FNFT_COMPLEX * const U,
```
This is minor, but `mat_mult_NxN` would be consistent with how the `poly_fmult` functions are named.

ShrinivasJC replied:
So should I move them to a new header `fnft__mat_mult.h`?

wahls replied:
I think it is fine if they stay in misc. These are simple helper functions.

ShrinivasJC replied:
Okay will only change the names then.

ShrinivasJC replied:
Should these be `fnft__misc_mat_mult_NxN` or just `fnft__mat_mult_NxN` ?

wahls replied:
`fnft__misc_mat_mult_NxN`

wahls commented on src/fnft_nsep.c:
```diff
@@ -112,28 +111,17 @@ INT fnft_nsep(const UINT D, COMPLEX const * const q,
-        const REAL eps_t = (T[1] - T[0])/(D - 1);
-        D_effective = D-1;
-        q_preprocessed = malloc(D_effective * sizeof(COMPLEX));
+        REAL Lam_shift = CARG(CEXP(I*phase_shift))/(-2*(T[1] - T[0]));
```
Could `CARG(CEXP(I*phase_shift))` not be replaced with `phase_shift`?

ShrinivasJC replied:
I am actually not sure about this. I had put that in incorrectly thinking that phase_shift needs to be between -PI and PI. However if you look at the reference it appears this need not be the case. Bijectivity of the  mapping from the phase_shift to Lam_shift  is lost if phase_shift is restricted to -PI to PI. I think the solution would be to indeed just replace CARG(CEXP(I*phase_shift)) with phase_shift and add a more elaborate note mentioning that it should be the unwrapped phase_shift. Please let me know your thoughts.

wahls replied:
I see. If you look down below, you indeed see that additional multiple's of pi mess up the factors. Maybe mentioned and add a check at the beginning of the function?

ShrinivasJC replied:
I did some checks in MATLAB and now realize that the additional multiples of pi will not mess things up because the `phase_shift` is used to obtain an intermediate signal which is also periodic. The `Lam_shift` is just relative to the spectrum of this intermediate signal.

ShrinivasJC replied:
Shall I mark this as resolved?

wahls replied:
Yes, please.

wahls commented on include/private/fnft__misc.h:
```diff
@@ -235,6 +236,66 @@ FNFT_UINT fnft__misc_nextpowerof2(const FNFT_UINT number);
+ * @param[in] x Real scalar value at which the value of the polynomial is to be calculated.
+ * @return Returns the value of nth degree Legendre polynomial at x.
+ */
+static inline FNFT_REAL fnft__misc_legendreP(const FNFT_UINT n, const FNFT_REAL x){
```
`misc_legendre_poly` would be more in line with other names

ShrinivasJC replied:
Changed.

wahls commented on include/private/fnft__nsev_slow.h:
```diff
@@ -0,0 +1,73 @@
+ *  defined in \link fnft_errwarn.h \endlink.
+ * @ingroup nse
+ */
+FNFT_INT fnft__nse_refine_roots_newton(const FNFT_UINT D,
```
Should this be renamed to `nsev_refine_bound_states_newton` since there are similar but different functions for the periodic NSE?

ShrinivasJC replied:
Changed.

PJ-Prins commented:
Unlike fnft_nsev, fnft_nsev_slow does not give the warning "FNFT Warning: Found more than *K_ptr bound states. Returning as many as possible." when applicable.

ShrinivasJC commented:
> Unlike fnft_nsev, fnft_nsev_slow does not give the warning "FNFT Warning: Found more than *K_ptr bound states. Returning as many as possible." when applicable.

fnft_nsev_slow does not yet have any method to find bound states on its own. It needs initial guesses. The user has to provide the guesses in the bound_states array and inform the routine about the number of bound state guesses in *K_ptr. It may so happen that two or more of the guesses converge to the same final value and are merged during filtering. Then the final number of bound states will always be lower and never higher. So I don't see where such a warning would be needed. Can you please quote the lines of the code like Sander has above?

wahls commented on src/private/fnft__misc.c:
```diff
@@ -320,3 +321,82 @@ UINT misc_nextpowerof2(const UINT number)
+        if (CABS(buf1[i]) > tmp)
+           tmp = CABS(buf1[i]);
+    if (CABS(buf1[D/2-1])/tmp > SQRT(EPSILON) || CABS(buf1[D/2+1])/tmp > SQRT(EPSILON))
+       WARN("Interplolation result maybe inaccurate. Signal may have discontinuity or insufficient samples provided.");
```
This warning could be improved. It is not clear what to do if the signal really is not bandlimited (i.e., the Fourier transform is decaying very slowly).

Maybe "Signal does not appear to be bandlimited. Interpolation step may be inaccurate. Try to reduce the step size, or switch to a discretization that does not require interpolation."? Is it clear from the documentation which discretizations require interpolation?

ShrinivasJC replied:
Currently it is not clear which discretizations require interpolation. I will add a list to the documentation. I will also then change the warning.

ShrinivasJC commented:
> Just for my understanding: Why is it not `CARG(q[D]/q[0])` if D-1 samples are passed? Has the number of samples in this test changed over time, or has some additional sample been added in the previous update? (This also applies to other tests.)

In the current release this test has 256 samples. I had added an extra sample which is the first sample of the next period. Now I have reverted to the same inputs as the current release version but I did not remove the extra sample. It is needed in the quasi-periodic cases to compute the phase shift. I can remove it from the periodic cases if it is confusing.

wahls commented on src/private/fnft__misc.c:
```diff
@@ -320,3 +321,82 @@ UINT misc_nextpowerof2(const UINT number)
+    for (i = 1; i<D; i++)
+        if (CABS(buf1[i]) > tmp)
+           tmp = CABS(buf1[i]);
+    if (CABS(buf1[D/2-1])/tmp > SQRT(EPSILON) || CABS(buf1[D/2+1])/tmp > SQRT(EPSILON))
```
This criterion seems to become less reliable as the number of samples increases.

Maybe we could improve it a bit. Let's assume we apply a perfect low-pass filter to q(t) at, say, 90% of the total bandwidth, leading to q_lp(t). 

I think we could compute the error ||q-q_lp||/||q||, which has a clear interpretation, directly in the frequency domain using Parseval's relation. We could warn if that is `>SQRT(EPSILON)`.

ShrinivasJC replied:
Indeed this seems to be a better criterion. I will implement it now.

wahls commented on CHANGELOG.md:
```diff
@@ -1,5 +1,21 @@
+### Changed
+
+- The routine fnft_nsep for the NFT of the periodic nonlinear Schroedinger equation now requires even number of samples and has an extra input phase shift over one quasi-period (see doc of fnft_nsep).
+
```
Improved computation of norming constants?

ShrinivasJC replied:
Added

PJ-Prins commented:
> fnft_nsev_slow does not yet have any method to find bound states on its own. It needs initial guesses. The user has to provide the guesses in the bound_states array and inform the routine about the number of bound state guesses in *K_ptr. It may so happen that two or more of the guesses converge to the same final value and are merged during filtering. Then the final number of bound states will always be lower and never higher. So I don't see where such a warning would be needed. Can you please quote the lines of the code like Sander has above?

I cannot refer to a line where a warning was _not_ printed...
I see your point now, but that does not become clear from the documentation. The documentation of fnft_nsev_slow says " Upon return, *K_ptr contains the number of actually detected bound states. Note that in order to skip the computation of the bound states completely, it is not sufficient to pass *K_ptr==0. Instead, one needs to pass bound_states==NULL." which suggests that if *K_ptr==0 and pass bound_states is not NULL, some kind of detection is taking place so upon return *K_ptr may contain a value larger than 0.

wahls commented on examples/fnft_nsev_slow_example.c:
```diff
@@ -0,0 +1,128 @@
+    // The types FNFT_UINT, FNFT_INT, FNFT_REAL and FNFT_COMPLEX are defined
+    // in the header fnft_numtypes.h
+
+    /** Step 2: Prepare calling fnft_nsev **/ 
```
`_slow` missing

ShrinivasJC replied:
Changed.

ShrinivasJC commented:
> > fnft_nsev_slow does not yet have any method to find bound states on its own. It needs initial guesses. The user has to provide the guesses in the bound_states array and inform the routine about the number of bound state guesses in *K_ptr. It may so happen that two or more of the guesses converge to the same final value and are merged during filtering. Then the final number of bound states will always be lower and never higher. So I don't see where such a warning would be needed. Can you please quote the lines of the code like Sander has above?
> 
> I cannot refer to a line where a warning was _not_ printed...
> I see your point now, but that does not become clear from the documentation. The documentation of fnft_nsev_slow says " Upon return, *K_ptr contains the number of actually detected bound states. Note that in order to skip the computation of the bound states completely, it is not sufficient to pass *K_ptr==0. Instead, one needs to pass bound_states==NULL." which suggests that if *K_ptr==0 and pass bound_states is not NULL, some kind of detection is taking place so upon return *K_ptr may contain a value larger than 0.

I see your point now but I can't think of how to improve it. Do you have any suggestion?

PJ-Prins commented:
> > > fnft_nsev_slow does not yet have any method to find bound states on its own. It needs initial guesses. The user has to provide the guesses in the bound_states array and inform the routine about the number of bound state guesses in *K_ptr. It may so happen that two or more of the guesses converge to the same final value and are merged during filtering. Then the final number of bound states will always be lower and never higher. So I don't see where such a warning would be needed. Can you please quote the lines of the code like Sander has above?
> > 
> > 
> > I cannot refer to a line where a warning was _not_ printed...
> > I see your point now, but that does not become clear from the documentation. The documentation of fnft_nsev_slow says " Upon return, *K_ptr contains the number of actually detected bound states. Note that in order to skip the computation of the bound states completely, it is not sufficient to pass *K_ptr==0. Instead, one needs to pass bound_states==NULL." which suggests that if *K_ptr==0 and pass bound_states is not NULL, some kind of detection is taking place so upon return *K_ptr may contain a value larger than 0.
> 
> I see your point now but I can't think of how to improve it. Do you have any suggestion?

The quick fix is to elaborate a bit more in the documentation of fnft_nsev_slow.
The underlying reason that it confuses me is that bound_state_localization contains in fact two independent options. One is how initial guesses are obtained ('fast eigenvalue' / 'subsampled fast eigenvalue' / 'user specified' (/ 'random' / ...)), the other is how the eigenvalues are refined after that ('No refinement' / 'Newton descent' ( / 'gradient descent'/ 'Nelder-Mead'/...)), and any combination is feasible. When you write in the documentation that only NEWTON (without link to the description of fnft_nsev_bsloc_NEWTON btw) is implemented, you actually mean to say that there is no algorithm implemented for the initial guesses, but that's not intuitively clear. The fundamental fix would imho be to split these two options.

ShrinivasJC commented:
> > > > fnft_nsev_slow does not yet have any method to find bound states on its own. It needs initial guesses. The user has to provide the guesses in the bound_states array and inform the routine about the number of bound state guesses in *K_ptr. It may so happen that two or more of the guesses converge to the same final value and are merged during filtering. Then the final number of bound states will always be lower and never higher. So I don't see where such a warning would be needed. Can you please quote the lines of the code like Sander has above?
> > > 
> > > 
> > > I cannot refer to a line where a warning was _not_ printed...
> > > I see your point now, but that does not become clear from the documentation. The documentation of fnft_nsev_slow says " Upon return, *K_ptr contains the number of actually detected bound states. Note that in order to skip the computation of the bound states completely, it is not sufficient to pass *K_ptr==0. Instead, one needs to pass bound_states==NULL." which suggests that if *K_ptr==0 and pass bound_states is not NULL, some kind of detection is taking place so upon return *K_ptr may contain a value larger than 0.
> > 
> > 
> > I see your point now but I can't think of how to improve it. Do you have any suggestion?
> 
> The quick fix is to elaborate a bit more in the documentation of fnft_nsev_slow.
> The underlying reason that it confuses me is that bound_state_localization contains in fact two independent options. One is how initial guesses are obtained ('fast eigenvalue' / 'subsampled fast eigenvalue' / 'user specified' (/ 'random' / ...)), the other is how the eigenvalues are refined after that ('No refinement' / 'Newton descent' ( / 'gradient descent'/ 'Nelder-Mead'/...)), and any combination is feasible. When you write in the documentation that only NEWTON (without link to the description of fnft_nsev_bsloc_NEWTON btw) is implemented, you actually mean to say that there is no algorithm implemented for the initial guesses, but that's not intuitively clear. The fundamental fix would imho be to split these two options.

I will add a link to fnft_nsev_bsloc_NEWTON but apart from that I am still unsure what additional documentation will be helpful.
@wahls What is your opinion on this?

wahls commented:
> > The quick fix is to elaborate a bit more in the documentation of fnft_nsev_slow.
> > The underlying reason that it confuses me is that bound_state_localization contains in fact two independent options. One is how initial guesses are obtained ('fast eigenvalue' / 'subsampled fast eigenvalue' / 'user specified' (/ 'random' / ...)), the other is how the eigenvalues are refined after that ('No refinement' / 'Newton descent' ( / 'gradient descent'/ 'Nelder-Mead'/...)), and any combination is feasible. When you write in the documentation that only NEWTON (without link to the description of fnft_nsev_bsloc_NEWTON btw) is implemented, you actually mean to say that there is no algorithm implemented for the initial guesses, but that's not intuitively clear. The fundamental fix would imho be to split these two options.
> 
> I will add a link to fnft_nsev_bsloc_NEWTON but apart from that I am still unsure what additional documentation will be helpful.
> @wahls What is your opinion on this?

I think the adding the link should suffice. The `bound_state_localization` option really refers to a complete localization procedure.

It might be beneficial to spit this option, but that's not the focus of this branch. Let's keep it in mind.

wahls commented on src/private/fnft__misc.c:
```diff
@@ -367,13 +367,20 @@ INT misc_resample(const UINT D, const REAL eps_t, COMPLEX const * const q,
+    // computation, the two norms of the 5% bandwidths on both ends of the
+    // spectrum are computed instead.
+    UINT Dlp = D/20;
+    REAL h = 0.001; //Arbitrary step-size required to ensure all the
```
Why is the provided step size `eps_t` not used instead?

ShrinivasJC replied:
Shouldn't matter but will change it.

wahls commented on src/private/fnft__misc.c:
```diff
@@ -367,13 +367,20 @@ INT misc_resample(const UINT D, const REAL eps_t, COMPLEX const * const q,
+    REAL h = 0.001; //Arbitrary step-size required to ensure all the
+                    // calculated norms are comparable.
+
+    REAL tmp = (SQRT(misc_l2norm2(Dlp, buf1+D/2-1-Dlp, 0, Dlp*h)) + 
```
I think the numerator should be a square root of the squared norms `(SQRT(l2norm2(...) + l2norm2(...))`, not a sum of the norms `SQRT(l2norm2(...))+SQRT(l2norm2(...))`.

ShrinivasJC replied:
Indeed that is a mistake. Fixing it now.

PJ-Prins commented:
> I will add a link to fnft_nsev_bsloc_NEWTON but apart from that I am still unsure what additional documentation will be helpful.

Instead of letting the reader infer what you need to say from things you do not write, write what you need to say. Suggestion:

[in,out] | K_ptr | Upon entry, *K_ptr should contain the length of the array bound_states. Upon return, *K_ptr contains the number of _**remaining bound states after refinement (see  fnft_nsev_bsloc_NEWTON of type fnft_nsev_bsloc_t)**_. Note that in order to skip the computation of the bound states completely, it is not sufficient to pass *K_ptr==0. Instead, one needs to pass bound_states==NULL.

[out] | bound_states | Array. Upon return, the routine has stored the detected bound states (aka eigenvalues) in the first *K_ptr entries of this array. If NULL is passed instead, the discrete spectrum will not be computed. Has to be preallocated by the user. The user can choose an arbitrary length. Currently _**no initial guess**_ method is supported hence bound_states should contain initial guesses of _**every bound state**_.

ShrinivasJC commented:
> > I will add a link to fnft_nsev_bsloc_NEWTON but apart from that I am still unsure what additional documentation will be helpful.
> 
> Instead of letting the reader infer what you need to say from things you do not write, write what you need to say. Suggestion:
> 
> [in,out] | K_ptr | Upon entry, *K_ptr should contain the length of the array bound_states. Upon return, *K_ptr contains the number of _**remaining bound states after refinement (see fnft_nsev_bsloc_NEWTON of type fnft_nsev_bsloc_t)**_. Note that in order to skip the computation of the bound states completely, it is not sufficient to pass *K_ptr==0. Instead, one needs to pass bound_states==NULL.
> 
> [out] | bound_states | Array. Upon return, the routine has stored the detected bound states (aka eigenvalues) in the first *K_ptr entries of this array. If NULL is passed instead, the discrete spectrum will not be computed. Has to be preallocated by the user. The user can choose an arbitrary length. Currently _**no initial guess**_ method is supported hence bound_states should contain initial guesses of _**every bound state**_.

Noted and changed.

wahls commented:
I am getting this warning with the latest version:
```
In file included from /home/sander/git/FNFT-higher_order/include/private/fnft__nse_discretization.h:33,
                 from /home/sander/git/FNFT-higher_order/include/private/fnft__nse_scatter.h:30,
                 from /home/sander/git/FNFT-higher_order/src/private/fnft__nse_scatter_bound_states.c:23:
/home/sander/git/FNFT-higher_order/src/private/fnft__nse_scatter_bound_states.c: In function ‘fnft__nse_scatter_bound_states’:
/home/sander/git/FNFT-higher_order/src/private/fnft__nse_scatter_bound_states.c:209:64: warning: implicit conversion from ‘fnft_nse_discretization_t’ to ‘fnft__akns_discretization_t’ [-Wenum-conversion]
  209 |         ret_code = akns_discretization_method_weights(&weights,discretization);
      |                                                                ^~~~~~~~~~~~~~
/home/sander/git/FNFT-higher_order/include/private/fnft__akns_discretization.h:171:90: note: in definition of macro ‘akns_discretization_method_weights’
  171 | #define akns_discretization_method_weights(...) fnft__akns_discretization_method_weights(__VA_ARGS__)
      |                                                                                          ^~~~~~~~~~~
```

ShrinivasJC commented:
> I am getting this warning with the latest version:
> 
> ```
> In file included from /home/sander/git/FNFT-higher_order/include/private/fnft__nse_discretization.h:33,
>                  from /home/sander/git/FNFT-higher_order/include/private/fnft__nse_scatter.h:30,
>                  from /home/sander/git/FNFT-higher_order/src/private/fnft__nse_scatter_bound_states.c:23:
> /home/sander/git/FNFT-higher_order/src/private/fnft__nse_scatter_bound_states.c: In function ‘fnft__nse_scatter_bound_states’:
> /home/sander/git/FNFT-higher_order/src/private/fnft__nse_scatter_bound_states.c:209:64: warning: implicit conversion from ‘fnft_nse_discretization_t’ to ‘fnft__akns_discretization_t’ [-Wenum-conversion]
>   209 |         ret_code = akns_discretization_method_weights(&weights,discretization);
>       |                                                                ^~~~~~~~~~~~~~
> /home/sander/git/FNFT-higher_order/include/private/fnft__akns_discretization.h:171:90: note: in definition of macro ‘akns_discretization_method_weights’
>   171 | #define akns_discretization_method_weights(...) fnft__akns_discretization_method_weights(__VA_ARGS__)
>       |                                                                                          ^~~~~~~~~~~
> ```

I was calling the incorrect akns routine instead of the nse one. Changed it. What options do you use while compiling? I seem to be missing these kinds of warnings.

ShrinivasJC commented:
Should `fnft_nsev_slow` have had an option which allows for computation of `a` and `aprime` for any complex `lambda`?

wahls commented on matlab/mex_fnft_nsep.c:
```diff
@@ -52,17 +55,20 @@ void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
     M = D;
     T = mxGetPr(prhs[1]);
-    kappa = (int)mxGetScalar(prhs[2]);
+    phase_shift = mxGetScalar(prhs[2]);
```
Could we make the `phase_shift` an optional parameter in the Matlab interface with default 0.0? (This version is per se fine, but it will break the Matlab examples in a paper that will be presented soon.)

ShrinivasJC replied:
Yes that can be done. Changing it now.

wahls commented on src/fnft_nsep.c:
```diff
@@ -198,15 +231,30 @@ static inline INT gridsearch(const UINT D,
     if (sheet_indices != NULL)
         return E_NOT_YET_IMPLEMENTED(sheet_indices, "Pass NULL");
+    
+    upsampling_factor = nse_discretization_upsampling_factor(opts_ptr->discretization);
```
Error check missing (signaled by `upsampling_factor == 0'), maybe also in other places.

ShrinivasJC replied:
Good catch!

ShrinivasJC replied:
Leaving this conversation open because it is possible I may have missed it in some place.

wahls commented on src/private/fnft__nsev_slow_testcases.c:
```diff
@@ -0,0 +1,664 @@
+#include <string.h> // for memcpy
+#endif
+
+INT nsev_slow_testcases(nsev_slow_testcases_t tc, const UINT D,
```
There seems to be a lot of overlap with `nsev_testcases`. Why not integrate these new test cases there instead (e.g., `nsev_testcases_SECH_FOCUSING2` instead of `nsev_slow_testcases_SECH_FOCUSING`)? That should reduce the overlap.

Also, only one `compare_nfs` function is needed I think.

`nsev_slow_testcases_test_fnft` could be renamed to `nsev_testcases_test_fnft_slow` and moved into nsev_testcases.c as well. This file could then be removed (that would also be consistent with the naming 'nsev_testcases`, which indicates testcases for the NSE with vanishing boundaries).

ShrinivasJC replied:
Hmm yes that can be done.

wahls commented on src/fnft_nsev_slow.c:
```diff
@@ -0,0 +1,683 @@
+ */
+
+
+// static inline INT refine_roots_newton(
```
remove dead comments

wahls commented on src/fnft_nsev_slow.c:
```diff
@@ -0,0 +1,683 @@
+
+// Auxiliary function for filtering: We assume that bound states must have
+// real part in the interval [-re_bound, re_bound].
+static inline REAL re_bound(const REAL eps_t)
```
Code duplication, `re_bound` is already defined in `fnft_sev.c`. `im_bound` as well. The code for computing the norming constants is also almost identical and could be merged.

`fnft_nsev(...)` and `fnft_nsev_slow(...)` and `fnft_nsev_base(...)` and `fnft_nsev_slow_base(...)` also have almost identical parts that could potentially factored out, even though that looks at bit more involved. Thoughts?

ShrinivasJC replied:
`re_bound`, `im_bound` and norming constants calculation could indeed be merged.  Although `fnft_nsev(...)` and `fnft_nsev_slow(...)` and `fnft_nsev_base(...)` and `fnft_nsev_slow_base(...)` have parts in common but there are significant enough differences that make me feel the code will get messy and difficult to read if merged.

wahls replied:
Ok, then I'd propose to do only the first three for now.

I'll have closer look at the rest next week. Maybe it would make sense to factor at least some parts out.

wahls replied:
I thought about how to deal with the duplicated code. It seems the easiest solution would be not to have a second NFT function (`fnft_nsev_slow`), but to integrate the slow part into `fnft_nsev` instead.

One could determine from the discretization and bound_state_location parameters whether the transfer matrix has to be build in `fnft_nsev_base`. The functions `tf2contspec`, `tf2boundstates` and `tf2normconsts_and_residues` could be extended to deal with slow discretizations as well.

Thoughts?

ShrinivasJC commented:
I have crudely merged the files. I will now see in detail what further reduction and streamlining can be done.

ShrinivasJC commented:
@wahls testing with valgrind using `--leak-check=full` and `--show-leak-kinds=all` options, `LEAK SUMMARY` shows some memory as `still reachable`. This is something within the FFT routines. Could you please have a look to see if this is important?

ShrinivasJC commented:
I looked again at the Richardson extrapolation code in `fnft_nsev`. There is very little code duplication in there. Moving parts of it to a new misc function feels unnecessary to me. Let me know what you think.

ShrinivasJC commented:
https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L555-L567

Should we add a limit based on `real max` as in Eq. 10 of https://arxiv.org/abs/2003.02215v2  ?

wahls commented:
> I looked again at the Richardson extrapolation code in `fnft_nsev`. There is very little code duplication in there. Moving parts of it to a new misc function feels unnecessary to me. Let me know what you think.

Looks very nice! Some small comments to follow later.

wahls commented:
> @wahls testing with valgrind using `--leak-check=full` and `--show-leak-kinds=all` options, `LEAK SUMMARY` shows some memory as `still reachable`. This is something within the FFT routines. Could you please have a look to see if this is important?

I think we have seen this before. If I am not mistaken, it is not a memory leak but intended:

> FFTW’s planner saves some other persistent data, such as the accumulated wisdom and a list of algorithms available in the
> current configuration. If you want to deallocate all of that and reset FFTW to the pristine state it was in when you started your 
> program, you can call:
>
> void fftw_cleanup(void);

http://www.fftw.org/doc/Using-Plans.html

Does this also occur with KissFFT?

wahls commented:
> https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L555-L567
> 
> Should we add a limit based on `real max` as in Eq. 10 of https://arxiv.org/abs/2003.02215v2 ?

Are we talking about the log(...)/T part in the minimum? I am not sure. What is the motivation behind it?

ShrinivasJC commented:
> > @wahls testing with valgrind using `--leak-check=full` and `--show-leak-kinds=all` options, `LEAK SUMMARY` shows some memory as `still reachable`. This is something within the FFT routines. Could you please have a look to see if this is important?
> 
> I think we have seen this before. If I am not mistaken, it is not a memory leak but intended:
> 
> > FFTW’s planner saves some other persistent data, such as the accumulated wisdom and a list of algorithms available in the
> > current configuration. If you want to deallocate all of that and reset FFTW to the pristine state it was in when you started your
> > program, you can call:
> > void fftw_cleanup(void);
> 
> http://www.fftw.org/doc/Using-Plans.html
> 
> Does this also occur with KissFFT?

You are right. It is seen only with FFTW and not with KissFFT.

ShrinivasJC commented:
> > https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L555-L567
> > 
> > Should we add a limit based on `real max` as in Eq. 10 of https://arxiv.org/abs/2003.02215v2 ?
> 
> Are we talking about the log(...)/T part in the minimum? I am not sure. What is the motivation behind it?

Yes the  log(...)/T part. While solving the Zakharov-Shabat system, we have `exp(-I*lambda*T)` as part of the initial condition. If `IMAG(lambda)>LOG(REAL_MAX)/T` then  `exp(-I*lambda*T)` is basically `INF`, so there is no meaning in trying to solve the ZS system.

wahls commented on src/fnft_nsev.c:
```diff
@@ -69,72 +63,99 @@ UINT fnft_nsev_max_K(const UINT D, fnft_nsev_opts_t const * const opts)
 /**
  * Fast nonlinear Fourier transform for the nonlinear Schroedinger
  * equation with vanishing boundary conditions.
  * See the header file for documentation.
```
I think it could be helpful to add a short comment about this specific function fits into the general scheme (takes care of Richardson and any required down/resampling before calling fnft_nsev_base).

wahls commented on src/fnft_nsev.c:
```diff
@@ -156,220 +177,366 @@ INT fnft_nsev(
-        goto release_mem;
+    
+    
+    upsampling_factor = nse_discretization_upsampling_factor(opts->discretization);
```
Short comment on why upsampling happens (some higher-order discretizations require non-equidistant grid)

wahls commented on src/fnft_nsev.c:
```diff
@@ -156,220 +177,366 @@ INT fnft_nsev(
-        ret_code = tf2contspec(deg, W, transfer_matrix, T, D, XI, M,
-            contspec, opts);
-        CHECK_RETCODE(ret_code, release_mem);
+    normconsts_or_residues_reserve = normconsts_or_residues;
```
Short comment on why this blob is here (will need buffer if Ricardson extrapolation is active?)

wahls commented on src/fnft_nsev.c:
```diff
@@ -156,220 +177,366 @@ INT fnft_nsev(
     
-    // Compute the discrete spectrum
-    if (kappa == +1 && bound_states != NULL) {
+    // Preprocess signal
```
This comment is not very useful. It is clear what `nse_preprocess_signal` preprocesses the signal. Maybe: What is happening during preprocessing? Why is it needed?

wahls commented on src/fnft_nsev.c:
```diff
@@ -156,220 +177,366 @@ INT fnft_nsev(
+            for (i=0; i<K_sub; i++)
+                bound_states_sub[i] = bound_states[i];
+        }
+        // Preparing q_preprocessed
```
Maybe mention here that in contrast to before, this q now is subsampled (to clarify why this is happening again).

wahls commented on src/fnft_nsev.c:
```diff
@@ -156,220 +177,366 @@ INT fnft_nsev(
+                }
+            }
+        }
+        if (kappa == +1 && bound_states != NULL && *K_ptr != 0 && K_sub != 0) {
```
Add comment "Richardson extrapolation of the discrete spectrum"

wahls commented on src/fnft_nsev.c:
```diff
@@ -156,220 +177,366 @@ INT fnft_nsev(
+        opts->bound_state_localization = bs_loc_opt;
+        opts->discspec_type = ds_type_opt;
+        
+        // Richardson step
```
Clarify to "Richardson extrapolation of the continuous spectrum"

wahls commented:
I really like this version. Two small issues:

1. I was wondering if the documentation of `fnft_nsev` should be updated to reflect that the computations may or may not be fast. Maybe something like "Nonlinear Fourier transform for the nonlinear Schroedinger equation with vanishing boundary conditions. Fast algorithms are used if the discretization supports it." Thoughts?

2. A formatting thing: I am seeing a lot of doubled empty lines where there should be only one. For example

https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L26

or

https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L219

Could that be caused by your editor using [Windows newlines](https://en.wikipedia.org/wiki/Newline#Representation)?

wahls commented:
> > > https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L555-L567
> > > 
> > > Should we add a limit based on `real max` as in Eq. 10 of https://arxiv.org/abs/2003.02215v2 ?
> > 
> > 
> > Are we talking about the log(...)/T part in the minimum? I am not sure. What is the motivation behind it?
> 
> Yes the log(...)/T part. While solving the Zakharov-Shabat system, we have `exp(-I*lambda*T)` as part of the initial condition. If `IMAG(lambda)>LOG(REAL_MAX)/T` then `exp(-I*lambda*T)` is basically `INF`, so there is no meaning in trying to solve the ZS system.

I am not convinced that this is a fundamental issue (in contrast to the argument used in `re_bound`). Since a(λ)→1 for |λ|→∞ in the upper half plane, I would expect that it can be taken into account. (Maybe split it up into D exponentials, and move these into the computation of the individual scattering matrices?) Or am I missing something here?

wahls commented on src/fnft_nsev.c:
```diff
@@ -604,37 +976,52 @@ static inline INT refine_roots_newton(
             bound_states[i] -= error;
             iter++;
-
             if (CIMAG(bound_states[i]) > im_bound_val
```
This might be inconsistent with the filtering method specified by the user. Should be pass the `bounding_box` that is computed anyway instead?

ShrinivasJC replied:
Indeed that would be more consistent. I will try to pass the `bounding_box`.

ShrinivasJC commented:
> I really like this version. Two small issues:
> 
> 1. I was wondering if the documentation of `fnft_nsev` should be updated to reflect that the computations may or may not be fast. Maybe something like "Nonlinear Fourier transform for the nonlinear Schroedinger equation with vanishing boundary conditions. Fast algorithms are used if the discretization supports it." Thoughts?
> 2. A formatting thing: I am seeing a lot of doubled empty lines where there should be only one. For example
> 
> https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L26
> 
> or
> 
> https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L219
> 
> Could that be caused by your editor using [Windows newlines](https://en.wikipedia.org/wiki/Newline#Representation)?

1. That is a good point. The information about each discretization is in `nse_discretization_t.h`. Should I just add a list of the discretizations which are "fast" and others "slow" in the documentation of `fnft_nsev' ?

2. Maybe. I will check.

wahls commented:
> > I really like this version. Two small issues:
> > 
> > 1. I was wondering if the documentation of `fnft_nsev` should be updated to reflect that the computations may or may not be fast. Maybe something like "Nonlinear Fourier transform for the nonlinear Schroedinger equation with vanishing boundary conditions. Fast algorithms are used if the discretization supports it." Thoughts?
> > 2. A formatting thing: I am seeing a lot of doubled empty lines where there should be only one. For example
> > 
> > https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L26
> > 
> > or
> > https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L219
> > 
> > Could that be caused by your editor using [Windows newlines](https://en.wikipedia.org/wiki/Newline#Representation)?
> 
>     1. That is a good point. The information about each discretization is in `nse_discretization_t.h`. Should I just add a list of the discretizations which are "fast" and others "slow" in the documentation of `fnft_nsev' ?

Yes, that would great.

>     2. Maybe. I will check.

Thank you!

ShrinivasJC commented:
> > > > https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L555-L567
> > > > 
> > > > Should we add a limit based on `real max` as in Eq. 10 of https://arxiv.org/abs/2003.02215v2 ?
> > > 
> > > 
> > > Are we talking about the log(...)/T part in the minimum? I am not sure. What is the motivation behind it?
> > 
> > 
> > Yes the log(...)/T part. While solving the Zakharov-Shabat system, we have `exp(-I*lambda*T)` as part of the initial condition. If `IMAG(lambda)>LOG(REAL_MAX)/T` then `exp(-I*lambda*T)` is basically `INF`, so there is no meaning in trying to solve the ZS system.
> 
> I am not convinced that this is a fundamental issue (in contrast to the argument used in `re_bound`). Since a(λ)→1 for |λ|→∞ in the upper half plane, I would expect that it can be taken into account. (Maybe split it up into D exponentials, and move these into the computation of the individual scattering matrices?) Or am I missing something here?

I unable to see how it can be split up into the D exponentials. Please provide some reference or we could discuss this at a later point as it is not crucial for the current pull-request.

wahls commented:
> > > > > https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L555-L567
> > > > > 
> > > > > Should we add a limit based on `real max` as in Eq. 10 of https://arxiv.org/abs/2003.02215v2 ?
> > > > 
> > > > 
> > > > Are we talking about the log(...)/T part in the minimum? I am not sure. What is the motivation behind it?
> > > 
> > > 
> > > Yes the log(...)/T part. While solving the Zakharov-Shabat system, we have `exp(-I*lambda*T)` as part of the initial condition. If `IMAG(lambda)>LOG(REAL_MAX)/T` then `exp(-I*lambda*T)` is basically `INF`, so there is no meaning in trying to solve the ZS system.
> > 
> > 
> > I am not convinced that this is a fundamental issue (in contrast to the argument used in `re_bound`). Since a(λ)→1 for |λ|→∞ in the upper half plane, I would expect that it can be taken into account. (Maybe split it up into D exponentials, and move these into the computation of the individual scattering matrices?) Or am I missing something here?
> 
> I unable to see how it can be split up into the D exponentials. Please provide some reference or we could discuss this at a later point as it is not crucial for the current pull-request.

Just `exp(I*lambda*T)=exp(I*(lambda/D + ... + lambda/D)*T)=exp(I*lambda*T/D) x ... x exp(I*lambda*T/D)`.

wahls commented:
I have no more comments. 

If there are no problems with the last few commits I made, I think we can start with the new release.

ShrinivasJC commented:
> > > > > > https://github.com/FastNFT/FNFT/blob/5eac55ebfbbe3fc574b207fb28dab95d2fca9b21/src/fnft_nsev.c#L555-L567
> > > > > > 
> > > > > > Should we add a limit based on `real max` as in Eq. 10 of https://arxiv.org/abs/2003.02215v2 ?
> > > > > 
> > > > > 
> > > > > Are we talking about the log(...)/T part in the minimum? I am not sure. What is the motivation behind it?
> > > > 
> > > > 
> > > > Yes the log(...)/T part. While solving the Zakharov-Shabat system, we have `exp(-I*lambda*T)` as part of the initial condition. If `IMAG(lambda)>LOG(REAL_MAX)/T` then `exp(-I*lambda*T)` is basically `INF`, so there is no meaning in trying to solve the ZS system.
> > > 
> > > 
> > > I am not convinced that this is a fundamental issue (in contrast to the argument used in `re_bound`). Since a(λ)→1 for |λ|→∞ in the upper half plane, I would expect that it can be taken into account. (Maybe split it up into D exponentials, and move these into the computation of the individual scattering matrices?) Or am I missing something here?
> > 
> > 
> > I unable to see how it can be split up into the D exponentials. Please provide some reference or we could discuss this at a later point as it is not crucial for the current pull-request.
> 
> Just `exp(I*lambda*T)=exp(I*(lambda/D + ... + lambda/D)*T)=exp(I*lambda*T/D) x ... x exp(I*lambda*T/D)`.

No matter how you calculate it, the quantity `exp(I*lambda*T)` itself will be greater than `realmax` for some combination of `lambda` and `T`. That is just a limitation of finite precision which I do not see a simply way of bypassing.

ShrinivasJC commented:
> I have no more comments.
> 
> If there are no problems with the last few commits I made, I think we can start with the new release.

Agreed. Please merge the pull request as I am not authorized to do so.

wahls commented:
higher order methods
